File: training-data/cpp/subset_sum.cpp
#include <iostream>
#include <vector>

bool hasSubsetSum(const std::vector<int>& nums, int target) {
    size_t n = nums.size();
    std::vector<std::vector<bool>> dp(n + 1, std::vector<bool>(target + 1, false));
    for (size_t i = 0; i <= n; i++) dp[i][0] = true;
    for (size_t i = 1; i <= n; i++) {
        for (int t = 1; t <= target; t++) {
            dp[i][t] = dp[i - 1][t];
            if (nums[i - 1] <= t) dp[i][t] = dp[i][t] || dp[i - 1][t - nums[i - 1]];
        }
    }
    return dp[n][target];
}

int main() {
    std::cout << hasSubsetSum({3, 34, 4, 12, 5, 2}, 9) << std::endl;
    return 0;
}

File: training-data/swift/subset_sum.swift
func hasSubsetSum(_ nums: [Int], target: Int) -> Bool {
    var dp = [Bool](repeating: false, count: target + 1)
    dp[0] = true

    for num in nums {
        var t = target
        while t >= num {
            if dp[t - num] {
                dp[t] = true
            }
            t -= 1
        }
    }
    return dp[target]
}

print(hasSubsetSum([3, 34, 4, 12, 5, 2], target: 9))  // true
print(hasSubsetSum([1, 2, 5], target: 4))             // false
